plugins/providers/gcloudiam/provider.go

```diff
@@ -75,11 +117,20 @@ func (p *Provider) GrantAccess(pc *domain.ProviderConfig, a *domain.Appeal) erro
 	}
 
 	if a.Resource.Type == ResourceTypeProject || a.Resource.Type == ResourceTypeOrganization {
-		if err := client.GrantAccess(a.AccountType, a.AccountID, a.Role); err != nil {
-			if errors.Is(err, ErrPermissionAlreadyExists) {
-				return nil
+		for _, rc := range pc.Resources {
+			for _, ro := range rc.Roles {
+				if a.Role == ro.ID {
+					for _, p := range ro.Permissions {
+						permission := fmt.Sprint(p)
+						if err := client.GrantAccess(a.AccountType, a.AccountID, permission); err != nil {
+							if !errors.Is(err, ErrPermissionAlreadyExists) {
+								return err
+							}
+						}
+					}
+					break
+				}
 			}
```

Comment:
Can you refactor this block? this is highly nested and difficult to understand.

Reply:
Refactored it to:

resourceRoleMap := make(map[string]*domain.Role)
for _, rc := range pc.Resources {
   for _, ro := range rc.Roles {
      resourceRole := fmt.Sprintf("%s-%s", rc.Type, ro.ID)
      resourceRoleMap[resourceRole] = ro
   }
}
appealResourceRole := fmt.Sprintf("%s-%s", a.Resource.Type, a.Role)
requestedRole := resourceRoleMap[appealResourceRole]

for _, p := range requestedRole.Permissions {
   permission := fmt.Sprint(p)
   if err := client.GrantAccess(a.AccountType, a.AccountID, permission); err != nil {
      if !errors.Is(err, ErrPermissionAlreadyExists) {
         return err
      }
   }
}